subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/impl/DirectorySnapshotter.java

```diff
@@ -270,7 +277,19 @@ protected FileVisitResult doPostVisitDirectory(Path dir, IOException exc) {
             if (isNotFileSystemLoopException(exc)) {
                 throw new UncheckedIOException(String.format("Could not read directory path '%s'.", dir), exc);
             }
-            builder.leaveDirectory();
+            boolean currentLevelComplete = builder.isCurrentLevelComplete();
```

Comment:
What does it mean that level is complete? So all files/directories on that level were visited?

Reply:
Oh I see, if there is a exclusion filter we don't visit some files/directories and we mark level as incomplete